kernel/src/path.rs

```diff
@@ -1,34 +1,108 @@
+    if path.is_empty() || path.ends_with('/') {
+        None
+    } else {
+        path.rsplit(DELIMITER).next()
+    }
+}
+
+fn get_version_opt(version_str_opt: Option<&str>, expected_digits: usize) -> Option<Version> {
+    version_str_opt.and_then(|version_str| {
+        if version_str.len() == expected_digits {
+            version_str.parse().ok()
+        } else {
+            None
+        }
+    })
+}
+
+pub(crate) fn version_from_location(location: &Url) -> Option<Version> {
+    let path = location.path();
+    get_filename(path)
+        .and_then(|f| f.split_once('.'))
+        .and_then(|(name, _)| get_version_opt(Some(name), 20))
+}
 
 impl<'a> LogPath<'a> {
+    pub(crate) fn new(url: &'a Url) -> Self {
+        let path = url.path();
+        let version_str = get_filename(path).and_then(|f| f.split_once('.'));
+        let version = version_str.and_then(|(name, _)| get_version_opt(Some(name), 20));
+
+        let mut is_commit = false;
+        let mut is_checkpoint = false;
+        let mut compacted_to_version = None;
+        if version.is_some() {
+            // could be a checkpoint or commit file, let's check
+            let (_, suffix) = version_str.unwrap(); // safe, version.is_some()
+            is_commit = suffix == "json"; // if we were just [version].json, we're a commit file
+
+            if !is_commit && suffix.starts_with("checkpoint.") {
+                let rest = &suffix[11..];
+                // check if name is just [version].checkpoint.parquet, i.e. we have a classic checkpoint
+                is_checkpoint = rest == "parquet";
+                if !is_checkpoint {
+                    // test if we're a multipart checkpoint
+                    let mut split = rest.splitn(3, '.');
+                    let (checkpoint_index, checkpoint_max, ext) = (
+                        get_version_opt(split.next(), 10),
+                        get_version_opt(split.next(), 10),
+                        split.next(),
+                    );
+                    is_checkpoint = checkpoint_index.is_some()
+                        && checkpoint_max.is_some()
+                        && ext == Some("parquet");
+                }
+            }
+
+            if !is_commit && !is_checkpoint {
+                // check if we're a compacted commit
+                if let Some((maybe_compacted_version, suffix)) = suffix.split_once('.') {
+                    if suffix == "json" {
+                        compacted_to_version = get_version_opt(Some(maybe_compacted_version), 20);
+                        is_commit = compacted_to_version.is_some()
+                    }
+                }
+            }
+        }
+        LogPath {
+            url,
+            path,
+            version,
+            _compacted_to_version: compacted_to_version,
+            is_commit,
+            is_checkpoint,
+        }
+    }
+
     pub(crate) fn child(&self, path: impl AsRef<str>) -> DeltaResult<Url> {
```

Comment:
This all seems like a lot of boilerplate for a purely crate internal struct. Maybe just make the fields all pub(crate) and remove the code?

Reply:
I'm assuming you mean all the accessors that don't do any work. Agree, I've removed them